File: frontend/src/app/core/services/analytics.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { AuthService } from './auth';

export interface ZoneDemandItem {
  zone: string;
  orders_count: number;
  pending_orders: number;
  planned_orders: number;
  delivered_orders: number;
  failed_orders: number;
  problematic_orders: number;
  total_kg: number;
  total_m3: number;
  average_kg: number;
  average_m3: number;
  urgent_orders: number;
  critical_orders: number;
  adr_orders: number;
  fragil_orders: number;
  perisabil_orders: number;
  strict_time_window_orders: number;
  unplanned_orders: number;
  zone_demand_score: number;
  demand_level: string;
  recommendation: string;
}

export interface ZoneDemandResponse {
  zones: ZoneDemandItem[];
}

export interface DriverZoneFitItem {
  driver_id: string;
  driver_name: string;
  total_stops_in_zone: number;
  completed_stops_in_zone: number;
  failed_stops_in_zone: number;
  skipped_stops_in_zone: number;
  on_time_rate: number;
  average_delay_minutes: number;
  average_service_minutes: number;
  minor_incidents: number;
  major_incidents: number;
  adr_experience_count: number;
  fragil_experience_count: number;
  perisabil_experience_count: number;
  critical_orders_handled: number;
  driver_zone_fit_score: number;
  recommendation_reason: string[];
}

export interface DriverZoneFitResponse {
  zone: string;
  recommended_drivers: DriverZoneFitItem[];
}

export interface VehicleTripFitItem {
  vehicle_id: string;
  plate: string;
  status: string;
  capacity_kg: number;
  capacity_m3: number;
  required_kg: number;
  required_m3: number;
  load_kg_percent: number;
  load_m3_percent: number;
  capacity_fit: boolean;
  recent_incidents: number;
  major_incidents: number;
  estimated_cost: number | null;
  vehicle_trip_fit_score: number;
  recommendation_reason: string[];
}

export interface VehicleTripFitResponse {
  recommended_vehicles?: VehicleTripFitItem[];
  error?: string;
}

export interface FleetUtilizationTripItem {
  trip_id: string;
  planned_date: string;
  vehicle_id: string;
  vehicle_plate: string;
  orders_count: number;
  stops_count: number;
  total_kg: number;
  total_m3: number;
  vehicle_capacity_kg: number;
  vehicle_capacity_m3: number;
  load_kg_percent: number;
  load_m3_percent: number;
  utilization_level: string;
  recommendation: string;
}

export interface FleetUtilizationSummary {
  used_vehicles: number;
  unused_vehicles: number;
  total_trips: number;
  capacity_compliant_trips: number;
  overloaded_trips: number;
  average_load_kg_percent: number;
  average_load_m3_percent: number;
  underutilized_trips: number;
  normal_trips: number;
  near_capacity_trips: number;
  efficient_trips: number;
}

export interface FleetUtilizationResponse {
  trips: FleetUtilizationTripItem[];
  summary: FleetUtilizationSummary;
}

export interface PlanningQualityResponse {
  planning_session_id: string;
  eligible_orders: number;
  planned_orders: number;
  unplanned_orders: number;
  planned_orders_rate: number;
  total_trips: number;
  total_stops: number;
  average_stops_per_trip: number;
  total_planned_km: number;
  total_planned_duration_min: number;
  warnings_count: number;
  critical_warnings_count: number;
  manual_adjustments_count: number;
  unassigned_trips_count: number;
  planning_quality_score: number;
  quality_level: string;
  recommendation: string;
  error?: string;
}

export interface UnplannedByZoneItem {
  zone: string;
  orders: number;
  kg: number;
  m3: number;
}

export interface UnplannedOrdersResponse {
  total_unplanned_orders: number;
  total_unplanned_kg: number;
  total_unplanned_m3: number;
  critical_unplanned_orders: number;
  urgent_unplanned_orders: number;
  by_reason: Record<string, number>;
  by_zone: UnplannedByZoneItem[];
  recommendations: string[];
}

export interface PlanVsActualGroupItem {
  name: string;
  on_time_rate: number;
  average_delay_minutes: number;
  average_service_minutes: number;
  stops_count: number;
}

export interface PlanVsActualResponse {
  planned_vs_actual_km_difference: number;
  planned_vs_actual_duration_difference: number;
  average_delay_minutes: number;
  max_delay_minutes: number;
  on_time_stops: number;
  late_stops: number;
  on_time_rate: number;
  average_real_service_time: number;
  average_service_time_deviation: number;
  trips_finished_on_time: number;
  trips_delayed: number;
  by_driver: PlanVsActualGroupItem[];
  by_vehicle: PlanVsActualGroupItem[];
  by_zone: PlanVsActualGroupItem[];
}

export interface ServiceTimeAccuracyItem {
  category: string;
  planned_service_minutes_avg: number;
  real_service_minutes_avg: number;
  deviation_minutes: number;
  deviation_percent: number;
  samples_count: number;
  recommendation: string;
}

export interface ServiceTimeAccuracyResponse {
  by_cargo_type: ServiceTimeAccuracyItem[];
  by_zone: ServiceTimeAccuracyItem[];
  by_stop_type: ServiceTimeAccuracyItem[];
}

export interface CostGroupItem {
  name: string;
  orders_count: number;
  trips_count: number;
  total_cost: number;
  cost_per_order: number;
  cost_per_km: number | null;
}

export interface ExpensiveTripItem {
  trip_id: string;
  actual_km: number;
  orders_count: number;
  total_cost: number;
  cost_per_order: number;
  recommendation: string;
}

export interface CostAnalyticsResponse {
  total_planned_cost: number;
  total_actual_cost: number;
  total_extra_cost: number;
  fuel_cost_planned: number;
  fuel_cost_actual: number;
  driver_cost_planned: number;
  driver_cost_actual: number;
  amortization_cost: number;
  cost_per_trip: number;
  cost_per_order: number;
  total_actual_km: number;
  cost_per_km: number | null;
  cost_per_kg: number | null;
  cost_per_m3: number | null;
  planned_vs_actual_variance: number;
  planned_vs_actual_variance_percent: number;
  by_driver: CostGroupItem[];
  by_vehicle: CostGroupItem[];
  by_zone: CostGroupItem[];
  top_expensive_trips: ExpensiveTripItem[];
}

export interface IncidentImpactItem {
  incident_id: string;
  original_trip_id: string;
  type: string;
  created_at: string;
  resolved_at: string | null;
  recovery_trip_id: string | null;
  affected_orders_count: number;
  affected_stops_count: number;
  affected_kg: number;
  affected_m3: number;
  extra_recovery_km: number;
  extra_recovery_duration_min: number;
  extra_recovery_cost: number;
  estimated_delay_minutes: number;
  incident_impact_score: number;
  impact_level: string;
  recommendation: string;
}

export interface IncidentImpactResponse {
  total_incidents: number;
  minor_incidents: number;
  major_incidents: number;
  interrupted_trips: number;
  recovery_trips_created: number;
  incident_rate: number;
  major_incident_rate: number;
  average_recovery_time_minutes: number | null;
  major_incidents_detail: IncidentImpactItem[];
}

@Injectable({
  providedIn: 'root',
})
export class AnalyticsService {
  private readonly apiUrl = 'http://127.0.0.1:8000/analytics';

  constructor(
    private http: HttpClient,
    private authService: AuthService,
  ) {}

  private buildHeaders() {
    return this.authService.getAuthHeaders();
  }

  private buildDateParams(startDate?: string, endDate?: string): HttpParams {
    let params = new HttpParams();
    if (startDate) {
      params = params.set('start_date', startDate);
    }
    if (endDate) {
      params = params.set('end_date', endDate);
    }
    return params;
  }

  getZoneDemand(startDate?: string, endDate?: string): Observable<ZoneDemandResponse> {
    return this.http.get<ZoneDemandResponse>(`${this.apiUrl}/zone-demand`, {
      headers: this.buildHeaders(),
      params: this.buildDateParams(startDate, endDate),
    });
  }

  getDriverZoneFit(zone: string, startDate?: string, endDate?: string): Observable<DriverZoneFitResponse> {
    let params = new HttpParams().set('zone', zone);
    if (startDate) {
      params = params.set('start_date', startDate);
    }
    if (endDate) {
      params = params.set('end_date', endDate);
    }

    return this.http.get<DriverZoneFitResponse>(`${this.apiUrl}/driver-zone-fit`, {
      headers: this.buildHeaders(),
      params,
    });
  }

  getVehicleTripFit(
    tripId?: string,
    orderIds?: string[],
    startDate?: string,
    endDate?: string,
  ): Observable<VehicleTripFitResponse> {
    let params = new HttpParams();
    if (tripId) {
      params = params.set('trip_id', tripId);
    }
    if (orderIds && orderIds.length > 0) {
      params = params.set('order_ids', orderIds.join(','));
    }
    if (startDate) {
      params = params.set('start_date', startDate);
    }
    if (endDate) {
      params = params.set('end_date', endDate);
    }

    return this.http.get<VehicleTripFitResponse>(`${this.apiUrl}/vehicle-trip-fit`, {
      headers: this.buildHeaders(),
      params,
    });
  }

  getFleetUtilization(startDate?: string, endDate?: string): Observable<FleetUtilizationResponse> {
    return this.http.get<FleetUtilizationResponse>(`${this.apiUrl}/fleet-utilization`, {
      headers: this.buildHeaders(),
      params: this.buildDateParams(startDate, endDate),
    });
  }

  getPlanningQuality(planningSessionId: string): Observable<PlanningQualityResponse> {
    return this.http.get<PlanningQualityResponse>(
      `${this.apiUrl}/planning-quality/${encodeURIComponent(planningSessionId)}`,
      {
        headers: this.buildHeaders(),
      },
    );
  }

  getUnplannedOrders(startDate?: string, endDate?: string): Observable<UnplannedOrdersResponse> {
    return this.http.get<UnplannedOrdersResponse>(`${this.apiUrl}/unplanned-orders`, {
      headers: this.buildHeaders(),
      params: this.buildDateParams(startDate, endDate),
    });
  }

  getPlanVsActual(startDate?: string, endDate?: string): Observable<PlanVsActualResponse> {
    return this.http.get<PlanVsActualResponse>(`${this.apiUrl}/plan-vs-actual`, {
      headers: this.buildHeaders(),
      params: this.buildDateParams(startDate, endDate),
    });
  }

  getServiceTimeAccuracy(startDate?: string, endDate?: string): Observable<ServiceTimeAccuracyResponse> {
    return this.http.get<ServiceTimeAccuracyResponse>(`${this.apiUrl}/service-time-accuracy`, {
      headers: this.buildHeaders(),
      params: this.buildDateParams(startDate, endDate),
    });
  }

  getCosts(startDate?: string, endDate?: string): Observable<CostAnalyticsResponse> {
    return this.http.get<CostAnalyticsResponse>(`${this.apiUrl}/costs`, {
      headers: this.buildHeaders(),
      params: this.buildDateParams(startDate, endDate),
    });
  }

  getIncidentImpact(startDate?: string, endDate?: string): Observable<IncidentImpactResponse> {
    return this.http.get<IncidentImpactResponse>(`${this.apiUrl}/incident-impact`, {
      headers: this.buildHeaders(),
      params: this.buildDateParams(startDate, endDate),
    });
  }

  getHealth(): Observable<{ status: string; module: string; endpoints: number }> {
    return this.http.get<{ status: string; module: string; endpoints: number }>(`${this.apiUrl}/health`, {
      headers: this.buildHeaders(),
    });
  }
}

File: backend/app/schemas/analytics.py
"""
Pydantic schemas pentru analytics endpoints.
"""
from pydantic import BaseModel
from datetime import datetime, date
from typing import Optional


# ============== Zone Demand ==============
class ZoneDemandItem(BaseModel):
    zone: str
    orders_count: int
    pending_orders: int
    planned_orders: int
    delivered_orders: int
    failed_orders: int
    problematic_orders: int
    total_kg: float
    total_m3: float
    average_kg: float
    average_m3: float
    urgent_orders: int
    critical_orders: int
    adr_orders: int
    fragil_orders: int
    perisabil_orders: int
    strict_time_window_orders: int
    unplanned_orders: int
    zone_demand_score: int
    demand_level: str  # LOW, MEDIUM, HIGH
    recommendation: str


class ZoneDemandResponse(BaseModel):
    zones: list[ZoneDemandItem]


# ============== Driver Zone Fit ==============
class DriverZoneFitItem(BaseModel):
    driver_id: str
    driver_name: str
    total_stops_in_zone: int
    completed_stops_in_zone: int
    failed_stops_in_zone: int
    skipped_stops_in_zone: int
    on_time_rate: float
    average_delay_minutes: float
    average_service_minutes: float
    minor_incidents: int
    major_incidents: int
    adr_experience_count: int
    fragil_experience_count: int
    perisabil_experience_count: int
    critical_orders_handled: int
    driver_zone_fit_score: int
    recommendation_reason: list[str]


class DriverZoneFitResponse(BaseModel):
    zone: str
    recommended_drivers: list[DriverZoneFitItem]


# ============== Vehicle Trip Fit ==============
class VehicleTripFitItem(BaseModel):
    vehicle_id: str
    plate: str
    status: str
    capacity_kg: float
    capacity_m3: float
    required_kg: float
    required_m3: float
    load_kg_percent: float
    load_m3_percent: float
    capacity_fit: bool
    recent_incidents: int
    major_incidents: int
    estimated_cost: Optional[float]
    vehicle_trip_fit_score: int
    recommendation_reason: list[str]


class VehicleTripFitResponse(BaseModel):
    recommended_vehicles: list[VehicleTripFitItem]


# ============== Fleet Utilization ==============
class FleetUtilizationTripItem(BaseModel):
    trip_id: str
    planned_date: str
    vehicle_id: str
    vehicle_plate: str
    orders_count: int
    stops_count: int
    total_kg: float
    total_m3: float
    vehicle_capacity_kg: float
    vehicle_capacity_m3: float
    load_kg_percent: float
    load_m3_percent: float
    utilization_level: str
    recommendation: str


class FleetUtilizationSummary(BaseModel):
    used_vehicles: int
    unused_vehicles: int
    total_trips: int
    capacity_compliant_trips: int
    overloaded_trips: int
    average_load_kg_percent: float
    average_load_m3_percent: float
    underutilized_trips: int
    near_capacity_trips: int
    efficient_trips: int


class FleetUtilizationResponse(BaseModel):
    trips: list[FleetUtilizationTripItem]
    summary: FleetUtilizationSummary


# ============== Planning Quality ==============
class PlanningQualityResponse(BaseModel):
    planning_session_id: str
    eligible_orders: int
    planned_orders: int
    unplanned_orders: int
    planned_orders_rate: float
    total_trips: int
    total_stops: int
    average_stops_per_trip: float
    total_planned_km: float
    total_planned_duration_min: int
    warnings_count: int
    critical_warnings_count: int
    manual_adjustments_count: int
    unassigned_trips_count: int
    planning_quality_score: int
    quality_level: str  # EXCELLENT, GOOD, MEDIUM, LOW
    recommendation: str


# ============== Unplanned Orders ==============
class UnplannedByZoneItem(BaseModel):
    zone: str
    orders: int
    kg: float
    m3: float


class UnplannedOrdersResponse(BaseModel):
    total_unplanned_orders: int
    total_unplanned_kg: float
    total_unplanned_m3: float
    critical_unplanned_orders: int
    urgent_unplanned_orders: int
    by_reason: dict[str, int]
    by_zone: list[UnplannedByZoneItem]
    recommendations: list[str]


# ============== Plan vs Actual ==============
class PlanVsActualGroupItem(BaseModel):
    name: str
    on_time_rate: float
    average_delay_minutes: float
    average_service_minutes: float
    stops_count: int


class PlanVsActualResponse(BaseModel):
    planned_vs_actual_km_difference: float
    planned_vs_actual_duration_difference: float
    average_delay_minutes: float
    max_delay_minutes: float
    on_time_stops: int
    late_stops: int
    on_time_rate: float
    average_real_service_time: float
    average_service_time_deviation: float
    trips_finished_on_time: int
    trips_delayed: int
    by_driver: list[PlanVsActualGroupItem]
    by_vehicle: list[PlanVsActualGroupItem]
    by_zone: list[PlanVsActualGroupItem]


# ============== Service Time Accuracy ==============
class ServiceTimeAccuracyItem(BaseModel):
    category: str
    planned_service_minutes_avg: float
    real_service_minutes_avg: float
    deviation_minutes: float
    deviation_percent: float
    samples_count: int
    recommendation: str


class ServiceTimeAccuracyResponse(BaseModel):
    by_cargo_type: list[ServiceTimeAccuracyItem]
    by_zone: list[ServiceTimeAccuracyItem]
    by_stop_type: list[ServiceTimeAccuracyItem]


# ============== Costs ==============
class CostGroupItem(BaseModel):
    name: str
    orders_count: int
    trips_count: int
    total_cost: float
    cost_per_order: float
    cost_per_km: Optional[float]


class ExpensiveTripItem(BaseModel):
    trip_id: str
    planned_km: float
    orders_count: int
    total_cost: float
    cost_per_order: float
    recommendation: str


class CostAnalyticsResponse(BaseModel):
    total_planned_cost: float
    total_actual_cost: float
    total_extra_cost: float
    fuel_cost_planned: float
    fuel_cost_actual: float
    driver_cost_planned: float
    driver_cost_actual: float
    amortization_cost: float
    cost_per_trip: float
    cost_per_order: float
    total_actual_km: float
    cost_per_km: Optional[float]
    cost_per_kg: Optional[float]
    cost_per_m3: Optional[float]
    planned_vs_actual_variance: float
    planned_vs_actual_variance_percent: float
    by_driver: list[CostGroupItem]
    by_vehicle: list[CostGroupItem]
    by_zone: list[CostGroupItem]
    top_expensive_trips: list[ExpensiveTripItem]


# ============== Incident Impact ==============
class IncidentImpactItem(BaseModel):
    incident_id: str
    original_trip_id: str
    type: str
    created_at: str
    resolved_at: Optional[str]
    recovery_trip_id: Optional[str]
    affected_orders_count: int
    affected_stops_count: int
    affected_kg: float
    affected_m3: float
    extra_recovery_km: float
    extra_recovery_duration_min: int
    extra_recovery_cost: float
    estimated_delay_minutes: int
    incident_impact_score: int
    impact_level: str  # LOW, MEDIUM, HIGH
    recommendation: str


class IncidentImpactResponse(BaseModel):
    total_incidents: int
    minor_incidents: int
    major_incidents: int
    interrupted_trips: int
    recovery_trips_created: int
    incident_rate: float
    major_incident_rate: float
    average_recovery_time_minutes: Optional[float]
    major_incidents_detail: list[IncidentImpactItem]
